### PR TITLE
TempyTable methods addition

### PR DESCRIPTION
**Pull request by nadaj**
Implements methods in issue: #41 
Adds example: #17

Methods mentioned in issue are added as well as an example for each method that is implemented.

### CODE REVIEW

**Comment by nadaj** [comment]
Will refactor and push once fixed.

File: tempy/widgets.py [hunk 1]
```diff
@@ -181,11 +182,78 @@ def make_footer(self, footer):
...
+    def col_class(self, css_class, col_index=None):
+        # adds css_class to every cell
+        if col_index is None:
+            for row in self.body.childs:
+                for cell in row.childs:
+                    cell.attr(klass=css_class)
+        elif col_index >= 0:
+            for row in self.body.childs:
+                if col_index < len(row.childs):
+                    row.childs[col_index].attr(klass=css_class)
```
Comment (Hrabal) [inline tempy/widgets.py]: At the end of this method, if no cell is affected (i.e: every row is shorter that the given col_index) i'd raise a WidgetError/IndexError

File: tempy/widgets.py [hunk 2]
```diff
@@ -181,11 +182,78 @@ def make_footer(self, footer):
...
+
+    def row_class(self, css_class, row_index=None):
+        # adds css_class to every row
+        if row_index is None:
+            for row in self.body.childs:
+                row.attr(klass=css_class)
+        elif row_index >= 0 and (row_index < len(self.body.childs)):
+            self.body.childs[row_index].attr(klass=css_class)
```
Comment (Hrabal) [inline tempy/widgets.py]: Same raise here and in the methods below.. maybe a _check_valid_index decorator or method can be implemented..

File: tempy/widgets.py [hunk 3]
```diff
@@ -181,11 +182,78 @@ def make_footer(self, footer):
...
+
+    def map_col(self, col_function, col_index=None):
+        # applies function to every cell
+        if col_index is None:
+            for row in self.body.childs:
```
Comment (Hrabal) [inline tempy/widgets.py]: What do you think about making an extra map_table method, called by both map_col and map_row in case no index is given?

**Comment by nadaj** [comment]
@Hrabal I agree with you on every issue, it's in the newest commit.

File: tempy/widgets.py [hunk 4]
```diff
@@ -181,11 +183,92 @@ def make_footer(self, footer):
...
+        elif self.is_row_within_bounds(row_index):
+            for cell in self.body.childs[row_index].childs:
+                if len(cell.childs) > 0:
+                    cell.childs[0] = row_function(cell.childs[0])
+
+    def map_table(self, format_function):
+        for row in self.body.childs:
+            for cell in row.childs:
+                if len(cell.childs) > 0:
+                    cell.childs[0] = format_function(cell.childs[0])
```
Comment (Hrabal) [inline tempy/widgets.py]: I just noticed that in those methods the function is applied to only the first child in the cell. Isn't better to apply the function to every element inside the cell? Don't know, never thought about that.. what you think?

Is it safe enough to assume that the user will apply functions to coherent cells? i.e: map_table will be used on a table in which every single cell contains a valid input for the applied function?
In the end it is the user's concern to check for the input of his functions.. In that case a `cell.childs = map(format_function , cell.childs)` would work better.

Also, what if one or more of this function calls breaks? Would it be useful to add a `ignore_errors` parameter so every function call is performed in a `try.. except` and exceptions are raised conditionally?

What you think?

Comment (nadaj) [reply tempy/widgets.py]: For the first issue you've mentioned I agree if you are thinking of putting multiple tags inside a table cell or type different than a string

For the exceptions part, are you thinking about putting that parameter in each of these newly implemented methods?

Comment (Hrabal) [reply tempy/widgets.py]: 1) Yeah I was thinking the situation in which a cell contains something more than a single string (multiple strings, other tag(s)). A cell is a TemPy tag so is can contain a list of anything...

2) Yes, I think the ignore_errors parameter should be in every newly implemented map_* method. It's a sort of "graceful function application" for those situation in which a formatting is needed but if something goes wrong better have a bad string than nothing..

Comment (nadaj) [reply tempy/widgets.py]: Is there any method implemented that set's content's text? I only see the text() and render() methods. I'm just asking so I don't implement and make duplicated code.

Comment (Hrabal) [reply tempy/widgets.py]: No there isn't, both render() and text() retrive the tag's content (with and without the html code). Insertion can be made with the APIs (text or anything) that in the end refers to DOMElement._insert(), there is no insertion specific for text though.

I thought about a string-specific method but that not seemed necessary, and yes seems a little tricky with nested ones..

**Comment by Hrabal** [comment]
Travis-CI is reporting a failure in tests, but running the tests locally don't raise anything so I'm merging.
Many thanks @nadaj!

**Comment by nadaj** [comment]
thanks @Hrabal! I know where the error in the test is. It's in the test_widgets.py in the method test_init_from_data_full. Once I added rows=15 in the TempyTable constructor, it was working and showing all passed. But I haven't pushed this since it wasn't the part of the issue. If you'd like I can push that fixed line.

**Comment by Hrabal** [comment]
No thanks, since the error was outside your PR I merged and worked on the bug myself, solved in the last commit.
It was a regression in TempyTable init, strange it never came up before!

**Pull request merged**